proposals/0127-get-sysvar-syscall.md

```diff
@@ -0,0 +1,180 @@
+---
+simd: '0127'
+title: sol_get_sysvar syscall
+authors:
+  - Richard Patel
+category: Standard
+type: Core
+status: Draft
+created: 2024-03-15
+---
+
+## Summary
+
+SIMD-0127 introduces the syscall `sol_get_sysvar` which allows BPF programs
+to retrieve any sysvar.
+
+New invariants are placed on the sysvar cache to enable safe retrieval.
+
+Existing syscalls to retrieve sysvars are henceforth considered obsolete.
+
+## Motivation
+
+Sysvars are special accounts that are implicitly modified by the runtime.
+Certain sysvars are essential for operation of the network.  For example, the
+system program reads the 'recent block hashes' sysvar to provide durable nonce
+functionality.
+
+Currently, programs can read sysvars can using one of these three methods:
+
+1. Via explicit account access by specifying the sysvar address in the list of
+   transaction accounts
+2. Via API calls to the sysvar cache (for native programs only)
+3. Via syscalls that copy the sysvar content to virtual machine memory
+   (for BPF programs only, does not work for all sysvars)
+
+Improvements to these APIs are motivated by the following reasons.
+
+### Allowing sysvar upgrades
+
+The function signature of a `sol_get_{...}_sysvar` syscall is as follows:
+
+```c
+uint64_t sol_get_example_sysvar( void * dest_addr );
+```
+
+When called, the syscall handler will copy the sysvar to the given memory
+location.  Notably, BPF programs do not inform the syscall handler of the
+expected size of the sysvar.  BPF programs thus effectively assume that the
+requested sysvars will forever have the same size.
+
+This prevents future upgrades that shrink or extend sysvars.
+
+### Syscall API fragmentation
+
+The following sysvars are currently accessible using syscalls:
+
+- `SysvarC1ock11111111111111111111111111111111` (via `sol_get_clock_sysvar`)
+- `SysvarEpochRewards1111111111111111111111111` (via `sol_get_epoch_rewards_sysvar`)
+- `SysvarEpochSchedu1e111111111111111111111111` (via `sol_get_epoch_schedule_sysvar`)
+- `SysvarFees111111111111111111111111111111111` (via `sol_get_fees_sysvar`)
+- `Sysvar1nstructions1111111111111111111111111` (indirectly via `sol_get_processed_sibling_instruction`)
+- `SysvarLastRestartS1ot1111111111111111111111` (via `sol_get_last_restart_slot`)
+- `SysvarRent111111111111111111111111111111111` (via `sol_get_rent_sysvar`)
+
+BPF programs cannot access the following sysvars using syscalls.
+
+- `SysvarRecentB1ockHashes11111111111111111111`
+- `SysvarRewards111111111111111111111111111111`
```

Comment:
This may or may not be necessary for the proposal, but these two sysvars are marked deprecated.

https://github.com/anza-xyz/agave/blob/e8526f60aa80d986111f935717f003d6be7670df/sdk/program/src/sysvar/recent_blockhashes.rs#L8

https://github.com/anza-xyz/agave/blob/e8526f60aa80d986111f935717f003d6be7670df/sdk/program/src/sysvar/rewards.rs#L1